app/src/main/java/com/topjohnwu/magisk/arch/BaseUIActivity.kt

```diff
@@ -54,17 +57,15 @@ abstract class BaseUIActivity<VM : BaseViewModel, Binding : ViewDataBinding> :
             .use { it.getDrawable(0) }
             .also { window.setBackgroundDrawable(it) }
 
-        window?.decorView?.let {
-            it.systemUiVisibility = (it.systemUiVisibility
-                    or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
-                    or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
-                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)
-        }
+        window.decorView.windowInsetsController?.hide(WindowInsets.Type.statusBars())
+        window.decorView.windowInsetsController?.hide(WindowInsets.Type.navigationBars())
+        window.decorView.windowInsetsController?.systemBarsBehavior = WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
+        window.setDecorFitsSystemWindows(false)
 
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
             window?.decorView?.post {
                 // If navigation bar is short enough (gesture navigation enabled), make it transparent
-                if (window.decorView.rootWindowInsets?.systemWindowInsetBottom ?: 0 < Resources.getSystem().displayMetrics.density * 40) {
+                if (window.decorView.rootWindowInsets?.isVisible(WindowInsets.Type.navigationBars()) == false) {
```

Comment:
The original purpose is to guess if gesture navigation is enabled, NOT if the navigation bar is visible. Your change is totally wrong.